zio-http/jvm/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala

```diff
@@ -106,38 +108,150 @@ private[netty] object NettyConnectionPool {
+        .handler(initializer)
+      localAddress.foreach(bootstrap.localAddress)
+
+      val channelFuture = bootstrap.connect()
+      val ch            = channelFuture.channel()
+      Scope.addFinalizer {
+        NettyFutureExecutor.executed {
+          channelFuture.cancel(true)
+          ch.close()
+        }.whenDiscard(ch.isOpen).ignoreLogged
+      }.uninterruptible *> NettyFutureExecutor.executed(channelFuture).as(ch)
+    }
+  }
+
+  /**
+   * Returns a sequence of connection attempts with their delays. Per RFC 8305,
+   * we start with IPv6, then after firstAddressFamilyDelay we try IPv4, then
+   * alternate between families.
+   */
+  private def sortAddresses(resolvedHosts: Chunk[InetAddress]): List[InetAddress] = {
+    val (ipv6Addresses, ipv4Addresses) = resolvedHosts.partition(_.isInstanceOf[Inet6Address])
+    val ipv6Iter                       = ipv6Addresses.iterator
+    val ipv4Iter                       = ipv4Addresses.iterator
+    val builder                        = List.newBuilder[InetAddress]
+    builder.sizeHint(resolvedHosts.size)
+
+    // Alternate between families
+    var useIpv6 = true
+    while (ipv6Iter.hasNext || ipv4Iter.hasNext) {
+
+      if (useIpv6 && ipv6Iter.hasNext) {
+        builder += ipv6Iter.next()
+      } else if (ipv4Iter.hasNext) {
+        builder += ipv4Iter.next()
+      } else if (ipv6Iter.hasNext) {
+        builder += ipv6Iter.next()
+      }
+
+      useIpv6 = !useIpv6
+    }
+
+    builder.result()
+  }
+
+  /**
+   * Implements Happy Eyeballs (RFC 8305) connection algorithm. Races connection
+   * attempts to IPv6 and IPv4 addresses with staggered delays.
+   */
+  private def happyEyeballsConnect(
+    resolvedHosts: Chunk[InetAddress],
+    channelFactory: JChannelFactory[JChannel],
+    eventLoopGroup: JEventLoopGroup,
+    location: URL.Location.Absolute,
+    initializer: ChannelInitializer[JChannel],
+    connectionTimeout: Option[Duration],
+    localAddress: Option[InetSocketAddress],
+  )(implicit trace: Trace): ZIO[Scope, Throwable, JChannel] = {
+
+    if (resolvedHosts.isEmpty) {
+      ZIO.fail(new RuntimeException("No addresses to connect to"))
+    } else if (resolvedHosts.size == 1) {
+      connectToAddress(
+        resolvedHosts.head,
+        channelFactory,
+        eventLoopGroup,
+        location,
+        initializer,
+        connectionTimeout,
+        localAddress,
+      )
+    } else {
+      val addresses = sortAddresses(resolvedHosts)
+      for {
+        lastFailed <- Queue.bounded[Unit](requestedCapacity = 1)
+        successful <- Ref.make(List.empty[JChannel])
+        _          <- ZIO.raceAll(
+          connectToAddress(
+            addresses.head,
+            channelFactory,
+            eventLoopGroup,
+            location,
+            initializer,
+            connectionTimeout,
+            localAddress,
+          ).onExit {
+            case e: Exit.Success[JChannel] => successful.update(channels => channels :+ e.value)
```

Comment:
I don't think this will work; Interruption might occur right between the moment that we create the channel and the evaluation of the next effect; before we reach this point.

We probably need to pass the ref into `connectToAddress` itself and add the channel to the list (potentially unsafely) at some uninterruptible point